hw/ip/csrng/doc/_index.md

```diff
@@ -91,3 +91,391 @@ This document aims to make the distinction between physical entropy and factory
+The application interface supports up to 15 hardware interfaces, and one software interface.
+
+A walk through of how CSNRG generates entropy bits begins with the application interface.
+An `instantiate` command is issued from one of the application interfaces.
+This request moves into the `cmd_stage` block.
+Here the request is arbitrated between all of the `cmd_stage` blocks.
+The winner will get its command moved into the command dispatch logic.
+A common state machine will process all application interface commands in order of arbitration.
+At this point, some seed entropy may be required depending on the command and any flags.
+If needed, a request to the entropy source hardware interface will be made.
+This step can take milliseconds if seed entropy is not immediately available.
+Once all of the prerequisites have been collected, a ctr_drbg command can be launched.
+This command will go into the ctr_drbg_cmd block.
+This ctr_drbg_cmd block uses two NIST-defined functions, the update and the block_encrypt functions.
+If the command is a generate, the ctr_drbg_cmd block will process the first half of the algorithm, and then pass it on to the ctr_drg_gen block.
+Additionally, the ctr_drbg_gen block also uses the update block and the block_encrypt block.
+To keep resources to a minimum, both of these blocks have arbiters to allow sharing between the ctr_drbg_cmd and ctr_drbg_gen blocks.
+The command field called `ccmd` (for current command) is sent along the pipeline to not only identify the command, but is also reused as a routing tag for the arbiters to use when returning the block response.
+
+Once the command has traversed through all of the ctr_drbg blocks, the result will eventually land into the state_db block.
+This block will hold the instance state for each application interface.
+The specific state information held in the instance is documented below.
+If the command was a `generate` command, the genbits data word will be returned to the requesting `cmd_stage` block.
+Finally, an `ack` response and status will be returned to the application interface once the command has been completely processed.
+
+
+## Block Diagram
+
+![CSRNG Block Diagram](csrng_blk_diag.svg)
+
+## Hardware Interfaces
+
+ {{< hwcfg "hw/ip/csrng/data/csrng.hjson" >}}
+
+
+## Design Details
+
+#### Non-blocking Commands
+Regarding command processing, all commands process immediately except for the generate command.
+The command generate length coutn (`glen`) is kept in the `cmd_stage` block.
+When the state_db block issues an ack to the `cmd_stage` block, the `cmd_stage` block increments an internal counter.
+This process repeats until the `glen` field value has been matched.
+Because each request is pipelined, requests from other `cmd_stage` blocks can be processed before the original generate command is completely done.
+This provides some interleaving of commands since a generate command can be programmed to take a very long time.
```

Comment:
ah is this idea here if a block is making use of `block_encrypt` and not `update`, the `update` function can be taken by another command? And we may end up processing a couple of commands in parallel?

Reply:
Yes...sort of. The whole flow is basically a pipeline, so yes, while a generate is processing, there are holes where another command can sneak in. I guess indeed parallel in that case.

Reply:
sounds great!